# Proyecto: Accidentes de tráfico

### Step 1: Problem definition

- Predicción de la gravedad de los accidentes con relación a factores ambientales, humanos y el estado de la vía.
- 


### Step 2: Acquiring and loading the data set

In [40]:
import kagglehub
from kagglehub import KaggleDatasetAdapter

import kagglehub

kagglehub.login()

In [41]:
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Load the latest version
data = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "oktayrdeki/traffic-accidents",
  "traffic_accidents.csv"
)

print("First 5 records:", data.head())

/tmp/ipykernel_8727/2972283283.py:5: DeprecationWarning: load_dataset is deprecated and will be removed in future version.
  data = kagglehub.load_dataset(


First 5 records:                crash_date traffic_control_device weather_condition  \
0  07/29/2023 01:00:00 PM         TRAFFIC SIGNAL             CLEAR   
1  08/13/2023 12:11:00 AM         TRAFFIC SIGNAL             CLEAR   
2  12/09/2021 10:30:00 AM         TRAFFIC SIGNAL             CLEAR   
3  08/09/2023 07:55:00 PM         TRAFFIC SIGNAL             CLEAR   
4  08/19/2023 02:55:00 PM         TRAFFIC SIGNAL             CLEAR   

       lighting_condition first_crash_type trafficway_type  \
0                DAYLIGHT          TURNING     NOT DIVIDED   
1  DARKNESS, LIGHTED ROAD          TURNING        FOUR WAY   
2                DAYLIGHT         REAR END  T-INTERSECTION   
3                DAYLIGHT            ANGLE        FOUR WAY   
4                DAYLIGHT         REAR END  T-INTERSECTION   

            alignment roadway_surface_cond road_defect  \
0  STRAIGHT AND LEVEL              UNKNOWN     UNKNOWN   
1  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
2  STRAIGHT AND

### Step 3: Store the information

Cambiar la forma de guardar los datos en la bd (hacerlo con script)

In [42]:
import sqlite3

#Creación de la db y tablas con datos

con = sqlite3.connect("../data/project_db.db")
data.to_sql('accident', con, if_exists='replace', index=False)

209306

### Step 4: Perform a descriptive analysis

In [43]:
data.describe()

num_units  injuries_total  injuries_fatal  injuries_incapacitating  \
count  209306.000000   209306.000000   209306.000000            209306.000000   
mean        2.063300        0.382717        0.001859                 0.038102   
std         0.396012        0.799720        0.047502                 0.233964   
min         1.000000        0.000000        0.000000                 0.000000   
25%         2.000000        0.000000        0.000000                 0.000000   
50%         2.000000        0.000000        0.000000                 0.000000   
75%         2.000000        1.000000        0.000000                 0.000000   
max        11.000000       21.000000        3.000000                 7.000000   

       injuries_non_incapacitating  injuries_reported_not_evident  \
count                209306.000000                  209306.000000   
mean                      0.221241                       0.121516   
std                       0.614960                       0.450865   
min                       0.000000                       0.000000   
25%                       0.000000                       0.000000   
50%                       0.000000                       0.000000   
75%                       0.000000                       0.000000   
max                      21.000000                      15.000000   

       injuries_no_indication     crash_hour  crash_day_of_week    crash_month  
count           209306.000000  209306.000000      209306.000000  209306.000000  
mean                 2.244002      13.373047           4.144024       6.771822  
std                  1.241175       5.603830           1.966864       3.427593  
min                  0.000000       0.000000           1.000000       1.000000  
25%                  2.000000       9.000000           2.000000       4.000000  
50%                  2.000000      14.000000           4.000000       7.000000  
75%                  3.000000      17.000000           6.000000      10.000000  
max                 49.000000      23.000000           7.000000      12.000000

### Step 5: Perform a full EDA

In [44]:
# Obtain dimensions
data.shape

(209306, 24)

In [45]:
# Obtain information about data types and non-null values
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209306 entries, 0 to 209305
Data columns (total 24 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   crash_date                     209306 non-null  object 
 1   traffic_control_device         209306 non-null  object 
 2   weather_condition              209306 non-null  object 
 3   lighting_condition             209306 non-null  object 
 4   first_crash_type               209306 non-null  object 
 5   trafficway_type                209306 non-null  object 
 6   alignment                      209306 non-null  object 
 7   roadway_surface_cond           209306 non-null  object 
 8   road_defect                    209306 non-null  object 
 9   crash_type                     209306 non-null  object 
 10  intersection_related_i         209306 non-null  object 
 11  damage                         209306 non-null  object 
 12  prim_contributory_cause       

- Hay un total de 209306 registros
- 10 variables son de tipo numérico, y 14 son categóricas.
- No se detectó ningún regristro con columnas nulas.


In [54]:
if data.duplicated().sum():
    print("Duplicados encontrados")
    total_data = data.drop_duplicates()
print(total_data.shape)
total_data.head()

Duplicados encontrados
(209275, 24)


crash_date traffic_control_device weather_condition  \
0  07/29/2023 01:00:00 PM         TRAFFIC SIGNAL             CLEAR   
1  08/13/2023 12:11:00 AM         TRAFFIC SIGNAL             CLEAR   
2  12/09/2021 10:30:00 AM         TRAFFIC SIGNAL             CLEAR   
3  08/09/2023 07:55:00 PM         TRAFFIC SIGNAL             CLEAR   
4  08/19/2023 02:55:00 PM         TRAFFIC SIGNAL             CLEAR   

       lighting_condition first_crash_type trafficway_type  \
0                DAYLIGHT          TURNING     NOT DIVIDED   
1  DARKNESS, LIGHTED ROAD          TURNING        FOUR WAY   
2                DAYLIGHT         REAR END  T-INTERSECTION   
3                DAYLIGHT            ANGLE        FOUR WAY   
4                DAYLIGHT         REAR END  T-INTERSECTION   

            alignment roadway_surface_cond road_defect  \
0  STRAIGHT AND LEVEL              UNKNOWN     UNKNOWN   
1  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
2  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
3  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
4  STRAIGHT AND LEVEL              UNKNOWN     UNKNOWN   

                         crash_type  ...        most_severe_injury  \
0            NO INJURY / DRIVE AWAY  ...   NO INDICATION OF INJURY   
1            NO INJURY / DRIVE AWAY  ...   NO INDICATION OF INJURY   
2            NO INJURY / DRIVE AWAY  ...   NO INDICATION OF INJURY   
3  INJURY AND / OR TOW DUE TO CRASH  ...  NONINCAPACITATING INJURY   
4            NO INJURY / DRIVE AWAY  ...   NO INDICATION OF INJURY   

  injuries_total injuries_fatal  injuries_incapacitating  \
0            0.0            0.0                      0.0   
1            0.0            0.0                      0.0   
2            0.0            0.0                      0.0   
3            5.0            0.0                      0.0   
4            0.0            0.0                      0.0   

  injuries_non_incapacitating  injuries_reported_not_evident  \
0                         0.0                            0.0   
1                         0.0                            0.0   
2                         0.0                            0.0   
3                         5.0                            0.0   
4                         0.0                            0.0   

   injuries_no_indication  crash_hour  crash_day_of_week  crash_month  
0                     3.0          13                  7            7  
1                     2.0           0                  1            8  
2                     3.0          10                  5           12  
3                     0.0          19                  4            8  
4                     3.0          14                  7            8  

[5 rows x 24 columns]

In [55]:
data.drop(["crash_date", "first_crash_type", "damage", "injuries_total"], axis = 1, inplace = True)
data.head()

traffic_control_device weather_condition      lighting_condition  \
0         TRAFFIC SIGNAL             CLEAR                DAYLIGHT   
1         TRAFFIC SIGNAL             CLEAR  DARKNESS, LIGHTED ROAD   
2         TRAFFIC SIGNAL             CLEAR                DAYLIGHT   
3         TRAFFIC SIGNAL             CLEAR                DAYLIGHT   
4         TRAFFIC SIGNAL             CLEAR                DAYLIGHT   

  trafficway_type           alignment roadway_surface_cond road_defect  \
0     NOT DIVIDED  STRAIGHT AND LEVEL              UNKNOWN     UNKNOWN   
1        FOUR WAY  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
2  T-INTERSECTION  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
3        FOUR WAY  STRAIGHT AND LEVEL                  DRY  NO DEFECTS   
4  T-INTERSECTION  STRAIGHT AND LEVEL              UNKNOWN     UNKNOWN   

                         crash_type intersection_related_i  \
0            NO INJURY / DRIVE AWAY                      Y   
1            NO INJURY / DRIVE AWAY                      Y   
2            NO INJURY / DRIVE AWAY                      Y   
3  INJURY AND / OR TOW DUE TO CRASH                      Y   
4            NO INJURY / DRIVE AWAY                      Y   

               prim_contributory_cause  num_units        most_severe_injury  \
0                  UNABLE TO DETERMINE          2   NO INDICATION OF INJURY   
1           IMPROPER TURNING/NO SIGNAL          2   NO INDICATION OF INJURY   
2                FOLLOWING TOO CLOSELY          3   NO INDICATION OF INJURY   
3                  UNABLE TO DETERMINE          2  NONINCAPACITATING INJURY   
4  DRIVING SKILLS/KNOWLEDGE/EXPERIENCE          2   NO INDICATION OF INJURY   

   injuries_fatal  injuries_incapacitating  injuries_non_incapacitating  \
0             0.0                      0.0                          0.0   
1             0.0                      0.0                          0.0   
2             0.0                      0.0                          0.0   
3             0.0                      0.0                          5.0   
4             0.0                      0.0                          0.0   

   injuries_reported_not_evident  injuries_no_indication  crash_hour  \
0                            0.0                     3.0          13   
1                            0.0                     2.0           0   
2                            0.0                     3.0          10   
3                            0.0                     0.0          19   
4                            0.0                     3.0          14   

   crash_day_of_week  crash_month  
0                  7            7  
1                  1            8  
2                  5           12  
3                  4            8  
4                  7            8

In [56]:
data.shape

(209306, 20)